docs/release-process-deep-dive.md

```diff
@@ -0,0 +1,596 @@
+      - uses: actions/upload-artifact@v4
+        with:
+          name: macos
+          if-no-files-found: error
+          retention-days: 7
+          path: |
+            dist/*.tar.gz
+            dist/*.zip
+            dist/*.pkg
+```
+</details>
+
+#### Building
+
+This job executes `script/release --local "$TAG_NAME" --platform macos` which uses `GoReleaser` to create the Go executables and `.zip` archives. See [how ./script/release works](#how-script-release-works) for further information.
+
+This job also executes `script/pkgmacos "$TAG_NAME"` to build a Universal (architecture independent) MacOS `.pkg` installer. See [how ./script/pkgmacos works](#how-script-pkgmacos-works) for further information.
+
+#### Signing
+
+For MacOS, the "signing" section refers to both signing and notarizing.
+
+There are three levels of "signing" that occur in this job:
+ * Signing of Go executables created by `GoReleaser` is performed in a [`GoReleaser` hook](https://github.com/cli/cli/blob/756f4ec04abdc9fdbab3fef35b182c546ef1dd17/.goreleaser.yml#L20).
+ * Notarization of `.zip` archives created by `GoReleaser` is performed by directly executing `script/sign dist/gh_*_macOS_*.zip`
+ * Signing of the `.pkg` installer in `./script/pkgmacos` when executing [`productbuild`](https://github.com/cli/cli/blob/1c74296d28cf5595008065d3ddf7061ca9388305/script/pkgmacos#L108). See warnings below.
+
+ > [!WARNING]
+> Although the job title is `Build & notarize universal macOS pkg installer`, the [`productbuild` docs](https://www.unix.com/man_page/osx/1/productbuild/) only refer to signing, thus notarization may not be the correct term here.
+
+> [!WARNING]
+> Although it appears as if signing the `.pkg` installer can occur if `inputs.environment == 'production'`, in practice, I don't believe we ever set `${{ vars.APPLE_DEVELOPER_INSTALLER_ID }}`, thus we always [skip signing](https://github.com/cli/cli/actions/runs/13271193192/job/37050749548#step:9:11).
+
+Signing of MacOS artifacts uses `codesign` and notarization uses `xcrun notarytool`, which submits the artifact to the Apple servers for additional checks.
+
+In order to perform signing, a keychain must be configured with the signing certificate.
+
+```sh
+keychain="$RUNNER_TEMP/buildagent.keychain"
+keychain_password="password1"
+
+security create-keychain -p "$keychain_password" "$keychain"
+security default-keychain -s "$keychain"
+security unlock-keychain -p "$keychain_password" "$keychain"
+
+base64 -D <<<"$APPLE_APPLICATION_CERT" > "$RUNNER_TEMP/cert.p12"
+security import "$RUNNER_TEMP/cert.p12" -k "$keychain" -P "$APPLE_APPLICATION_CERT_PASSWORD" -T /usr/bin/codesign
+security set-key-partition-list -S "apple-tool:,apple:,codesign:" -s -k "$keychain_password" "$keychain"
+rm "$RUNNER_TEMP/cert.p12"
+```
+
+> [!NOTE]
+> TODO: Further Documentation required on these lines.
+
+> [!NOTE]
+> TODO: Further Documentation required on the difference between signing and notarization.
```

Comment:
I keep asking myself the same question so appreciate the note

For more information, https://developer.apple.com/documentation/security/notarizing-macos-software-before-distribution